asterius/src/Asterius/Types.hs

```diff
@@ -94,6 +95,32 @@ data AsteriusStatics
       }
   deriving (Show, Data)
 
+-- ----------------------------------------------------------------------------
+
+data AsteriusCachedModule
```

Comment:
I suggest adding `AsteriusModule` as a field of `AsteriusCachedModule`, instead of just duplicating all the fields.

Reply:
Since `gcSections` calls `collectEntitySymbols` on each `AsteriusStatics` and `Function` within a module, I basically paired each `SymbolSet` with the entity it comes from:
```diff
- staticsMap :: SymbolMap AsteriusStatics,
+ staticsMap :: SymbolMap (AsteriusStatics, SymbolSet),
---
- functionMap :: SymbolMap Function,
+ functionMap :: SymbolMap (Function, SymbolSet),
```
If I understand correctly `gcSections`, the `SymbolSet` for a complete module depends on the `export_funcs`, which are passed dynamically. This means that we cannot precompute `SymbolSet` at the module-level statically. Hence the representation I use does not reuse `AsteriusModule`. Did I miss something? Do you think we can do better than this?

Reply:
My suggestion is having an extra `SymbolMap SymbolSet` field in addition to the `AsteriusModule` field. This represents the mapping from a single symbol to a set of symbols, which is the "adjacency list" representation of a directed graph. Thus we can achieve minimal intrusion to the original logic.

Reply:
Hmmmm, I see what you mean, I'll introduce an `dependencyMap :: SymbolMap SymbolSet` then. FWIW, I dislike it a little that this introduces an implicit invariant that
```haskell
keys staticsMap `union` keys functionMap == keys dependencyMap`
```
but you are right that it is the least intrusive.